content/knowledge-base/guides/write-a-composition-function-in-go.md

```diff
@@ -0,0 +1,825 @@
+create an Amazon Web Services (AWS) S3 bucket for each entry in the names array.
+<!-- vale gitlab.FutureTense = NO -->
+
+To write a function in Go you:
+
+1. Install the tools you need to write the function
+1. Initialize the function from a template
+1. Edit the template to add the function's logic
+1. Test the function end-to-end
+1. Build and push the function to a package repository
+
+This guide covers each of these steps in detail.
+
+## Install the tools you need to write the function
+
+To write a function in Go you need:
+
+* [Go](https://go.dev/dl/) 1.21 or newer. The guide uses Go 1.21.
+* [Docker Engine](https://docs.docker.com/engine/). This guide uses Engine v24.
+* The [Crossplane CLI](https://github.com/crossplane/docs/pull/584) v1.14 or
+  newer. This guide uses Crossplane CLI v1.14.
+
+You don't need access to a Kubernetes cluster or a Crossplane control plane to
+build or test a composition function.
+
+## Initialize the function from a template
+
+Use the Crossplane CLI's `crossplane beta xpkg init` command to initialize a new
+function named function-xbuckets. When you run this command the CLI
+initializes your function using the
+[function-template-go](https://github.com/crossplane/function-template-go)
+GitHub repository as a template.
+
+```shell
+crossplane beta xpkg init function-xbuckets function-template-go -d function-xbuckets
+Initialized package "function-xbuckets" in directory "/home/negz/control/negz/function-xbuckets" from https://github.com/crossplane/function-template-go/tree/91a1a5eed21964ff98966d72cc6db6f089ad63f4 (main)
+```
+
+The `crossplane beta init xpkg` command creates a directory named
+`function-xbuckets`. When you run the command the new directory should look like
+this:
+
+```shell
+ls function-xbuckets
+Dockerfile  fn.go  fn_test.go  go.mod  go.sum  input/  LICENSE  main.go  package/  README.md  renovate.json
+```
+
+The `fn.go` file is where you add the function's code. It's useful to know about
+some other files in the template:
+
+* `main.go` runs the function. You don't need to edit `main.go`.
+* `Dockerfile` builds the function runtime. You don't need to edit `Dockerfile`.
+* The `input` directory defines the function's input type.
+* The `package` directory contains metadata used to build the function package.
+
+{{<hint "tip">}}
+<!-- vale gitlab.FutureTense = NO -->
+<!--
+This tip talks about future plans for Crossplane.
+-->
+In v1.14 of the Crossplane CLI `crossplane beta xpkg init` just clones a
+template GitHub repository. In a future release the command will automate tasks
+like replacing the template name with the new function's name. See Crossplane
+issue [#4941](https://github.com/crossplane/crossplane/issues/4941) for details.
+<!-- vale gitlab.FutureTense = YES -->
+{{</hint>}}
+
+You must make some changes before you start adding code:
+
+* Edit `package/crossplane.yaml` to change the package's name.
+* Edit `go.mod` to change the Go module's name.
```

Comment:
does this also require the user to manually update package import statements throughout the code base? or is that done automatically?

should they run `go mod tidy` after this step?

Reply:
I'm intentionally glossing over this. There's only one package import statement that uses the import path, and it's to import the input types. That will go away when the code that imports the input types is deleted/replaced.

Reply:
Also, pretty sure they don't need to run `go mod tidy` after renaming the module (the module shouldn't appear in its own dependencies).